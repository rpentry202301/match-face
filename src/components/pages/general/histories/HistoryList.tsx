"use client";

import GrayButton from "@/components/ui/button/GrayButton";
import Link from "next/link";
import { memo, useEffect, useState } from "react";

interface HistoryListProps {
  click?: boolean;
  month?: string;
  skill?: number[] | undefined;
}
interface TypeProject {
  administratorId: number;
  answered: boolean;
  comment: { context: string };
  createdAt: string;
  createdUser: string;
  deadline: string;
  id: number;
  project: { id: number; name: string; detail: string };
  questionList: [];
  requestAt: string;
  updateAt: string;
  updateUser: string;
  user: {};
}

const HistoryList: React.FC<HistoryListProps> = memo(
  ({ click, month, skill }) => {
    const [selectProject, setSelectProject] = useState<TypeProject[]>([]);

    useEffect(() => {
      async function getSelectedData() {
        const userId = 1;
        const response = await fetch("http://localhost:3000/api/histories", {
          cache: "no-store",
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            user_id: userId,
            month: month,
            skill: skill,
          }),
        });
        if (!response.ok) throw new Error("Failed to fetch data");
        const FetchData = await response.json();
        const selectProject = FetchData.answerRequestList;
        setSelectProject(selectProject);
        // console.log("取得したデータ", selectProject, month, skill);
      }
      getSelectedData();
    }, [click]);

    // 詳細の文字数制限
    const truncateString = (str: any, num: any) => {
      return str.length <= num ? str : str.slice(0, num) + "...";
    };
    // 回答日の表示
    const formedDeadline = (inputData: string): string => {
      const date = new Date(inputData);
      const year = date.getFullYear();
      const month = date.getMonth() + 1;
      const day = date.getDate();
      return `${year}年${month < 10 ? "0" : ""}${month}月${
        day < 10 ? "0" : ""
      }${day}日`;
    };

    return (
      <div>
        <table className="table-auto border border-collapse my-20 w-[80vw]">
          <thead>
            <tr>
              <th className="border p-4 bg-neutral-300">コメント</th>
              <th className="border  bg-neutral-300">回答日</th>
              <th className="border  bg-neutral-300">案件名</th>
              <th className="border  bg-neutral-300">案件概要</th>
              <th className="border  bg-neutral-300"></th>
            </tr>
          </thead>
          <tbody>
            {selectProject.map((project) => (
              <tr key={project.id}>
                <td className="border text-center p-3">
                  {project.comment.context.length > 0 ? (
                    <span className=" bg-orange text-white p-2 w-40 h-auto w-15 rounded py-1 px-2 text-xs">
                      新着
                    </span>
                  ) : (
                    ""
                  )}
                </td>
                <td className="border text-center p-3">
                  {" "}
                  {formedDeadline(project.updateAt)}
                </td>
                <td className="border text-center px-4">
                  {project.project.name}
                </td>

                <td
                  className="border text-center px-4 "
                  data-test={`projectDetail${project.id}`}
                >
                  {truncateString(project.project.detail, 30)}
                </td>
                <td className="border text-center px-4">
                  <Link href={`result/${project.project.id}`}>
                    <GrayButton
                      label={"詳細"}
                      className="w-15 rounded py-1 px-2 text-xs"
                      value={project.project.id}
                    />
                  </Link>
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    );
  }
);

HistoryList.displayName = "HistoryList";
export default HistoryList;
