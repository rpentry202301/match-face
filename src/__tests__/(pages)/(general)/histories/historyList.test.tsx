import { render, screen, cleanup, act, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import { mockHistoriesData, skillData } from "./mock";
import HistoriesPage from "@/app/(pages)/(general)/(other)/histories/page";
import HistoryList from "@/components/pages/general/histories/HistoryList";
import userEvent from "@testing-library/user-event";
import OrangeButton from "@/components/ui/button/OrangeButton";

// スナップショットテスト
describe("スナップショットテスト", () => {
  global.fetch = jest.fn().mockImplementation((url, config) => {
    if (url === "http://localhost:3000/api/histories/skills") {
      return {
        ok: true,
        json: async () => skillData,
      };
    }
    if (url === "http://localhost:3000/api/histories") {
      return {
        ok: true,
        json: async () => mockHistoriesData,
      };
    }
  });
  it("スナップショット", async () => {
    const { container } = render(<HistoryList />);
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
});

// 機能・インタラクションテスト
describe("一般ユーザー質問一覧画面", () => {
  global.fetch = jest.fn().mockImplementation((url, config) => {
    if (url === "http://localhost:3000/api/histories/skills") {
      return {
        ok: true,
        json: async () => skillData,
      };
    }
    if (url === "http://localhost:3000/api/histories") {
      return {
        ok: true,
        json: async () => mockHistoriesData,
      };
    }
  });
  describe("HistoryListコンポーネント", () => {
    it("取得データの要素数の<button>が存在する", async () => {
      await waitFor(() => {
        render(<HistoryList />);
      });
      await waitFor(() => {
        const element = screen.getAllByRole("button");
        screen.debug();
        expect(element).toHaveLength(2);
      });
    });
  });

  // describe("HistoryListに引数を与える", () => {
  //   it("取得データの要素数は０である", async () => {
  //     await render(<HistoryList month={"2023-10-01"} />);
  //     await waitFor(() => {
  //       const element = screen.getAllByRole("button");
  //       screen.debug();
  //       expect(element).toHaveLength(0);
  //     });
  //   });
  // });
  //   describe("ボタン", () => {
  //     it("回答済みであれば、ボタンの色が緑である&&「確認する」", async () => {
  //       let element;
  //       await waitFor(() => {
  //         element = screen.getByTestId(
  //           `confirmButton${mockData.answerRequestList[0].id}`
  //         );
  //       });
  //       console.log(element);
  //       expect(element).toHaveClass("bg-green");
  //       expect(element).toHaveTextContent("確認する");
  //     });
  //     it("確認するボタンにresult/[id]へのLinkがある", async () => {
  //       let element;
  //       await waitFor(() => {
  //         element = screen.getByTestId(
  //           `linkButton${mockData.answerRequestList[0].id}`
  //         );
  //       });
  //       screen.debug();
  //       expect(element).toHaveAttribute(
  //         "href",
  //         `result/${mockData.answerRequestList[0].id}`
  //       );
  //     });
  //   });
});
