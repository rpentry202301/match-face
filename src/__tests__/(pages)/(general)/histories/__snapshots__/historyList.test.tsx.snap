// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`スナップショットテスト スナップショット 1`] = `
<div>
  <div>
    <table
      class="table-auto border border-collapse my-20 w-[80vw]"
    >
      <thead>
        <tr>
          <th
            class="border p-4 bg-neutral-300"
          >
            コメント
          </th>
          <th
            class="border  bg-neutral-300"
          >
            回答日
          </th>
          <th
            class="border  bg-neutral-300"
          >
            案件名
          </th>
          <th
            class="border  bg-neutral-300"
          >
            案件概要
          </th>
          <th
            class="border  bg-neutral-300"
          />
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;
